elab-util-extra/src/Language/Reflection/Unify/Interface.idr

```diff
@@ -0,0 +1,274 @@
+  showPrec p t =
+    showCon p "MkUniTask" $
+      joinBy "" $
+        [ showArg t.lfv
+        , showArg t.lhsFreeVars
+        , showArg t.lhsExpr
+        , showArg t.rfv
+        , showArg t.rhsFreeVars
+        , showArg t.rhsExpr
+        ]
+
+||| Free variable output data
+public export
+record FVData where
+  constructor MkFVData
+  ||| Free variable name
+  name : Name
+  ||| Free variable hole name
+  holeName : String
+  ||| Free variable count
+  rig : Count
+  ||| Free variable PiInfo
+  piInfo : PiInfo TTImp
+  ||| Free variable type
+  type : TTImp
+  ||| Free variable value
+  value : Maybe TTImp
+
+%name FVData fv, fvData
+
+%runElab derive "FVData" [Show, Eq]
+
+export
+Interpolation FVData where
+  interpolate (MkFVData n h r p t v) = joinBy "" [ showPiInfo p $ showCount r "\{n} \{h} : \{show t}", " = \{show v}" ]
+
+||| Make FVData out of most its components and an argument
+export
+makeFVData : (String, Arg, Name, TTImp, Maybe TTImp) -> FVData
+makeFVData (h, fv, n, t, v) = MkFVData n h fv.count fv.piInfo t v
+
+public export
+record FVDeps (freeVars : Nat) where
+  constructor MkFVDeps
+  typeDeps : FinSet freeVars
+  valueDeps : FinSet freeVars
+  piInfoDeps : FinSet freeVars
+
+%name FVDeps deps
+
+{freeVars : Nat} -> Show (FVDeps freeVars) where
+  showPrec p t =
+    showCon p "MkFVDeps" $
+      joinBy "" $
+        [ showArg t.typeDeps
+        , showArg t.valueDeps
+        , showArg t.piInfoDeps
+        ]
+
+{freeVars : Nat} -> Eq (FVDeps freeVars) where
+  (==) a b = a.typeDeps == b.typeDeps && a.valueDeps == b.valueDeps && a.piInfoDeps == b.piInfoDeps
+
+export
+mergeDeps : FVDeps fv -> FinSet fv
+mergeDeps (MkFVDeps typeDeps valueDeps piInfoDeps) = union typeDeps $ union valueDeps piInfoDeps
+
+export
+removeDeps : FinSet fv -> FVDeps fv -> FVDeps fv
+removeDeps d =
+  { typeDeps $= flip difference d
+  , valueDeps $= flip difference d
+  , piInfoDeps $= flip difference d
+  }
+
+||| Free variable depenfdency graph
+public export
+record DependencyGraph where
+  constructor MkDG
+  ||| Free variable amount
+  freeVars : Nat
+  ||| Free variable data
+  fvData : Vect freeVars FVData
+  ||| Free variable dependency matrix
+  fvDeps : Vect freeVars $ FVDeps freeVars
+  ||| The set of all i: (Fin freeVars), where (index i fvData).value = None
+  empties : FinSet freeVars
+  ||| For all i : (Fin freeVars); (lookup (index i fvData).name nameToId) = i
+  nameToId : SortedMap Name $ Fin freeVars
+  ||| For all i : (Fin freeVars); (lookup (index i fvData).holeName holeToId) = i
+  holeToId : SortedMap String $ Fin freeVars
+
+%name DependencyGraph dg, depGraph
+
+%runElab derive "DependencyGraph" [Show]
+
+export
+Eq DependencyGraph where
+  (==) (MkDG a b c d e f) (MkDG a' b' c' d' e' f') with (decEq a a')
+   (==) (MkDG a b c d e f) (MkDG a' b' c' d' e' f') | Yes p =
+    a == a' && b == (rewrite p in b') && c == (rewrite p in c') &&
+      d == (rewrite p in d') && e == (rewrite p in e') && f == (rewrite p in f')
```

Comment:
You should match `Yes p` on `Yes Refl`, replace `a'` to `a` in the pattern-matching LHS, and thus get rid of all `rewrite`s